exonum/src/messages2/authorization.rs

```diff
@@ -0,0 +1,147 @@
+/// | `32`     | message class          |
+/// | `33`     | message type           |
+/// | `34..N`  | Payload                |
+/// | `N..N+64`| Signature                |
+///
+///
+/// Every creation of `SignedMessage` lead to signature verification, or data signing procedure,
+/// which can slowdown your code. Beware `SignedMessage` message, this procedure is not free.
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub struct SignedMessage {
+    pub(in messages) raw: Vec<u8>,
+}
+
+impl SignedMessage {
+    pub(crate) fn new(
+        cls: u8,
+        tag: u8,
+        value: Vec<u8>,
+        author: PublicKey,
+        secret_key: &SecretKey,
+    ) -> SignedMessage {
+        let mut buffer = Vec::new();
+        let signature = Self::sign(&value, secret_key).expect("Couldn't form signature");
+        buffer.extend_from_slice(author.as_ref());
+        buffer.push(cls);
+        buffer.push(tag);
+        buffer.extend_from_slice(value.as_ref());
+        buffer.extend_from_slice(signature.as_ref());
+        SignedMessage { raw: buffer }
+    }
+
+    /// Create `SignedMessage` wrapper from raw buffer.
+    /// Checks binary format and signature.
+    pub fn verify_buffer(buffer: Vec<u8>) -> Result<Self, Error> {
+        if buffer.len() <= EMPTY_SIGNED_MESSAGE_SIZE {
+            bail!("Message too short message_len = {}", buffer.len())
+        }
+        let signed = SignedMessage { raw: buffer };
+
+        {
+            let pk = signed.author();
+            let signature = signed.signature();
+            let payload = signed.payload();
+
+            Self::verify(payload, &signature, &pk)?;
+        }
+
+        Ok(signed)
+    }
+
+    #[cfg(test)]
+    pub(crate) fn unchecked_from_vec(buffer: Vec<u8>) -> Self {
+        SignedMessage { raw: buffer }
+    }
+    #[cfg(not(test))]
+    pub(in messages) fn unchecked_from_vec(buffer: Vec<u8>) -> Self {
+        SignedMessage { raw: buffer }
+    }
+
+    #[allow(unsafe_code)]
+    pub(in messages) fn author(&self) -> PublicKey {
+        PublicKey::from_slice(&self.raw[0..PUBLIC_KEY_LENGTH]).expect("Couldn't read PublicKey")
+    }
+
+    pub(in messages) fn message_class(&self) -> u8 {
+        self.raw[PUBLIC_KEY_LENGTH]
+    }
+
+    pub(in messages) fn message_type(&self) -> u8 {
+        self.raw[PUBLIC_KEY_LENGTH + 1]
+    }
+
+    pub(in messages) fn payload(&self) -> &[u8] {
+        let sign_idx = self.raw.len() - SIGNATURE_LENGTH;
+        &self.raw[PUBLIC_KEY_LENGTH + 2..sign_idx]
+    }
+
+    #[allow(unsafe_code)]
+    pub(in messages) fn signature(&self) -> Signature {
+        let sign_idx = self.raw.len() - SIGNATURE_LENGTH;
+        Signature::from_slice(&self.raw[sign_idx..]).expect("Couldn't read signature")
+    }
+
+    /// Return byte array representation of internal data.
+    pub fn raw(&self) -> &[u8] {
+        &self.raw
+    }
+
+    pub fn hash(&self) -> Hash {
+        hash(self.raw())
```

Comment:
`&self.raw` ?
The reason is if we for example will remove method `pub fn raw(&self) -> &[u8]` we will break `pub fn hash(&self) -> Hash` as well.

Reply:
Using a getter method has its merits too. If we change the implementation details and store messages in a more convoluted way then we may not have to manually update all the impls which use the getter as it will still return the necessary raw message bytes.